### ISSUE TITLE
Mypy crashes with an `INTERNAL ERROR` on my Meson refactoring branch

### ISSUE THREAD

**Issue by mensinda** [python/mypy#11192]
**Crash Report**

I am working on some Meson refactorings and I am currently getting a mypy `INTERNAL ERROR`.

**Traceback**

```
Running mypy (this can take some time) ...
mesonbuild/interpreterbase/interpreterbase.py:59: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.920+dev.3ec0284513359810bb5b149f94eec236d2e8da6c
Traceback (most recent call last):
  File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/__main__.py", line 23, in <module>
    console_entry()
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/__main__.py", line 11, in console_entry
    main(None, sys.stdout, sys.stderr)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/main.py", line 87, in main
    res, messages, blockers = run_build(sources, options, fscache, t0, stdout, stderr)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/main.py", line 165, in run_build
    res = build.build(sources, options, None, flush_errors, fscache, stdout, stderr)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/build.py", line 179, in build
    result = _build(
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/build.py", line 254, in _build
    graph = dispatch(sources, manager, stdout)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/build.py", line 2707, in dispatch
    process_graph(graph, manager)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/build.py", line 3031, in process_graph
    process_stale_scc(graph, scc, manager)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/build.py", line 3123, in process_stale_scc
    mypy.semanal_main.semantic_analysis_for_scc(graph, scc, manager.errors)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/semanal_main.py", line 78, in semantic_analysis_for_scc
    process_top_levels(graph, scc, patches)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/semanal_main.py", line 199, in process_top_levels
    deferred, incomplete, progress = semantic_analyze_target(next_id, state,
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/semanal_main.py", line 326, in semantic_analyze_target
    analyzer.refresh_partial(refresh_node,
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/semanal.py", line 402, in refresh_partial
    self.refresh_top_level(node)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/semanal.py", line 413, in refresh_top_level
    self.accept(d)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/semanal.py", line 4991, in accept
    node.accept(self)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/nodes.py", line 1083, in accept
    return visitor.visit_assignment_stmt(self)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/semanal.py", line 2005, in visit_assignment_stmt
    s.rvalue.accept(self)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/nodes.py", line 1646, in accept
    return visitor.visit_index_expr(self)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/semanal.py", line 3927, in visit_index_expr
    self.analyze_type_application(expr)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/semanal.py", line 3933, in analyze_type_application
    types = self.analyze_type_application_args(expr)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/semanal.py", line 3985, in analyze_type_application_args
    analyzed = self.anal_type(typearg, allow_unbound_tvars=True,
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/semanal.py", line 5086, in anal_type
    typ = typ.accept(a)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/types.py", line 511, in accept
    return visitor.visit_unbound_type(self)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/typeanal.py", line 163, in visit_unbound_type
    typ = self.visit_unbound_type_nonoptional(t, defining_literal)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/typeanal.py", line 231, in visit_unbound_type_nonoptional
    special = self.try_analyze_special_unbound_type(t, fullname)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/typeanal.py", line 320, in try_analyze_special_unbound_type
    return self.analyze_callable_type(t)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/typeanal.py", line 756, in analyze_callable_type
    return ret.accept(self)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/types.py", line 1112, in accept
    return visitor.visit_callable_type(self)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/typeanal.py", line 545, in visit_callable_type
    variables=self.anal_var_defs(variables),
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/typeanal.py", line 986, in anal_var_defs
    return [self.anal_var_def(vd) for vd in var_defs]
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/typeanal.py", line 986, in <listcomp>
    return [self.anal_var_def(vd) for vd in var_defs]
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/typeanal.py", line 978, in anal_var_def
    var_def.upper_bound.accept(self),
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/types.py", line 1922, in accept
    return visitor.visit_placeholder_type(self)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/typeanal.py", line 669, in visit_placeholder_type
    n = None if t.fullname is None else self.api.lookup_fully_qualified(t.fullname)
  File "/home/daniel/.local/lib/python3.9/site-packages/mypy/semanal.py", line 4377, in lookup_fully_qualified
    next_sym = n.names[parts[i]]
KeyError: 'interpreterbase'
mesonbuild/interpreterbase/interpreterbase.py:59: : note: use --pdb to drop into pdb

```

**To Reproduce**

Source code: https://github.com/mensinda/meson/tree/mypyError

Sadly, I am not able to narrow this down so I guess my current WIP branch of Meson has to do. Also, Meson is only partially typed, so we launch mypy through our `run_mypy.py` in the root of the repo. Finally, this error occurs after some not finished refactoring with probably broken type annotations.

**Your Environment**

- Mypy version used: `0.920+dev.3ec0284513359810bb5b149f94eec236d2e8da6c`
- Mypy command-line flags: See https://github.com/mensinda/meson/blob/mypyError/run_mypy.py
- Mypy configuration options from `mypy.ini` (and other config files): See https://github.com/mensinda/meson/tree/mypyError
- Python version used: 3.9.7
- Operating system and version: Arch Linux

**Comment by Akuli** [comment]
I reduced this to a more minimal example.

```
mesonbuild/
├── __init__.py                  Empty file
└── interpreterbase
    ├── baseobjects.py           Content below
    ├── __init__.py              Contains one line: from . import interpreterbase
    └── interpreterbase.py       Content below
```

`mesonbuild/interpreterbase/baseobjects.py` contains:
```python
from typing import Union, Dict, TypeVar

TYPE_var = Union[int, "ThisClassDoesNotExist"]
TYPE_HoldableTypes = Dict[str, TYPE_var]
InterpreterObjectTypeVar = TypeVar(
    "InterpreterObjectTypeVar", bound=TYPE_HoldableTypes
)

def get_file_internal():
    from ..interpreterbase import FeatureNew
```

`mesonbuild/interpreterbase/interpreterbase.py` contains:
```python
from .baseobjects import InterpreterObjectTypeVar, NonExistent

from typing import Dict, Callable

HolderMapType = Dict[
    int,
    Callable[[], NonExistent[InterpreterObjectTypeVar]]
]
```

output:
```
$ mypy mesonbuild/interpreterbase/interpreterbase.py 
mesonbuild/interpreterbase/interpreterbase.py:1: error: Module "mesonbuild.interpreterbase.baseobjects" has no attribute "NonExistent"
mesonbuild/interpreterbase/interpreterbase.py:5: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
If this issue continues with mypy master, please report a bug at https://github.com/python/mypy/issues
version: 0.920+dev.3ec0284513359810bb5b149f94eec236d2e8da6c
mesonbuild/interpreterbase/interpreterbase.py:5: : note: please use --show-traceback to print a traceback when reporting a bug
```

**Comment by Michael0x2a** [comment]
This no longer repros on master -- it looks like the crash was fixed by #11205, according to git-bisect.
